### PR TITLE
Allow slices with negative step in C

### PR DESCRIPTION
**Pull request by saniQQ**
- Use ternary operator to handle slices with negative `step` when `allow_negative_index` decorator is used;
- Fix duplicated operator in case of literal values;
- Add tests.

### CODE REVIEW

File: pyccel/codegen/printing/ccode.py [hunk 1]
```diff
@@ -702,13 +702,13 @@ def _new_slice_with_processed_arguments(_slice, array_size, allow_negative_index
         # negative start and end in slice
         if isinstance(start, PyccelUnarySub) and isinstance(start.args[0], LiteralInteger):
             start = PyccelMinus(array_size, start.args[0])
-        elif allow_negative_index and not isinstance(start, LiteralInteger):
+        elif allow_negative_index and not isinstance(start, LiteralInteger) and not isinstance(start, PyccelArraySize):
```
Comment (MenoIy) [inline pyccel/codegen/printing/ccode.py:705]: this can be :
```
isinstance(start, (LiteralInteger, PyArraySize))
```

File: pyccel/codegen/printing/ccode.py [hunk 2]
```diff
@@ -702,13 +702,13 @@ def _new_slice_with_processed_arguments(_slice, array_size, allow_negative_index
...
             start = IfTernaryOperator(PyccelLt(start, LiteralInteger(0)),
                             PyccelMinus(array_size, start), start)
 
         if isinstance(stop, PyccelUnarySub) and isinstance(stop.args[0], LiteralInteger):
             stop = PyccelMinus(array_size, stop.args[0])
-        elif allow_negative_index and not isinstance(stop, LiteralInteger):
+        elif allow_negative_index and not isinstance(stop, LiteralInteger) and not isinstance(stop, PyccelArraySize):
```
Comment (MenoIy) [inline pyccel/codegen/printing/ccode.py:711]: here too

**Review by MenoIy — CHANGES_REQUESTED** [review]

File: pyccel/codegen/printing/ccode.py (hunk 1 above)
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py]: will fix

**Review by MenoIy — APPROVED** [review]

File: pyccel/codegen/printing/ccode.py [hunk 3]
```diff
@@ -702,13 +702,13 @@ def _new_slice_with_processed_arguments(_slice, array_size, allow_negative_index
         # negative start and end in slice
         if isinstance(start, PyccelUnarySub) and isinstance(start.args[0], LiteralInteger):
             start = PyccelMinus(array_size, start.args[0])
-        elif allow_negative_index and not isinstance(start, LiteralInteger):
+        elif allow_negative_index and not isinstance(start, (LiteralInteger, PyccelArraySize)):
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:705]: Why is this needed? What happens if I do:
```
@types('int[:]')
def f(a):
    import numpy as np
    s = np.shape(a)[0]
    b = np.zeros(s+1)
    print(b[s])
```
?

File: pyccel/codegen/printing/ccode.py [hunk 4]
```diff
@@ -720,12 +720,12 @@ def _new_slice_with_processed_arguments(_slice, array_size, allow_negative_index
...
         elif isinstance(step, PyccelUnarySub) and isinstance(step.args[0], LiteralInteger):
-            start = array_size if _slice.start is None else start
+            start = PyccelMinus(array_size, LiteralInteger(1)) if _slice.start is None else start
             stop = LiteralInteger(0) if _slice.stop is None else stop
 
         # variable step in slice
         elif allow_negative_index and step and not isinstance(step, LiteralInteger):
-            start = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)), start, stop)
+            start = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)), start, PyccelMinus(stop, LiteralInteger(1)))
             stop = IfTernaryOperator(PyccelGt(step, LiteralInteger(0)), stop, start)
```
Comment (EmilyBourne) [inline pyccel/codegen/printing/ccode.py:729]: Has this been tested? I just want to check that there isn't a problem due to the fact that `stop` is constructed using the modified `start`

File: pyccel/stdlib/ndarrays/ndarrays.h [hunk 5]
```diff
@@ -74,7 +74,7 @@ typedef struct  s_ndarray
     int32_t                 length;
     /* size of the array */
     int32_t                 buffer_size;
-    /*  */
+    /* True if the array do not own the data */
```
Comment (EmilyBourne) [inline pyccel/stdlib/ndarrays/ndarrays.h:77]: Grammar
`do`-> `does`

File: pyccel/codegen/printing/ccode.py (hunk 3 above)
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:705]: that condition will occur only in handling a slice object and because we know that an array size will always be positive we don't need to have to add a ternary operator

File: pyccel/codegen/printing/ccode.py (hunk 4 above)
Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:729]: yes it is tested this code :
```python
@allow_negative_index('a')
def ok():
    a = np.array([1, 2, 3, 4, 5])
    j = -1
    b = a[::j]
    for j in range(5):
        print(b[j])
```

is translated to :
```C
void ok(void)
{
    t_ndarray a;
    int64_t j;
    t_ndarray b;

    a = array_create(1, (int32_t[]){5}, nd_int64);
    int64_t array_dummy_0001[] = {1, 2, 3, 4, 5};
    memcpy(a.nd_int64, array_dummy_0001, a.buffer_size);

    j = -1;
    b = array_slicing(a, new_slice((j > 0) ? 0 : a.shape[0] - 1, (j > 0) ? a.shape[0] : (j > 0) ? 0 : a.shape[0] - 1, j));
    for (j = 0; j < 5; j += 1)
    {
        printf("%ld\n", b.nd_int64[get_index(b, j)]);
    }
    free_array(a);
    free_pointer(b);
}
```
and there is a test for this case in the tests that i added

Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py:729]: Is there really no way for this to be simpler?!
```
new_slice((j > 0) ? 0 : a.shape[0] - 1, (j > 0) ? a.shape[0] : (j > 0) ? 0 : a.shape[0] - 1, j)
```

Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:729]: i can't think of a more simpler way because the `stop` depends on the `start` and both depend on `j` which is **dynamic**.

Comment (yguclu) [reply pyccel/codegen/printing/ccode.py:729]: ### Potential bug?

I am a bit confused by the second argument, where you check two times for the same condition `y > 0`:
```c
(j > 0) ? a.shape[0] : (j > 0) ? 0 : a.shape[0] - 1
```
- If `j > 0` you enter the first case, that is `a.shape[0]`;
- If `j <= 0` you enter the second case, where you check again if `j > 0`... but this is false and hence you always enter the case `a.shape[0] - 1`... But I think we should print `0` instead!

In summary I think you should print
```c
(j > 0) ? a.shape[0] - 1 : 0
```
and the final expression should be
```c
new_slice((j > 0) ? 0 : a.shape[0] - 1, (j > 0) ? a.shape[0] - 1: 0, j)
```
Am I missing something here? Please let me know.

### Cleaner print

It seems to me that this expression is a bit clearer:
```c
(j > 0) ? new_slice(0, a.shape[0] - 1, j) : new_slice(a.shape[0] - 1, 0, j)
```
where a TernaryOperator node has two Slice nodes as arguments. What do you think?

Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py:729]: yeah there was a bug i didn't notice it, i fixed it.
to make the `TernaryOperator` accept a `Slice`  as argument the `Slice` object needs to have a (`rank`, `shape`, `precision`, `dtype`)
if we want to add them to the `Slice` object , i could make the (`rank`, `shape`, `precision`) default to `None`, but i don't know what to make the `dtype`

Comment (EmilyBourne) [reply pyccel/codegen/printing/ccode.py]: I think it would make sense for `Slice` to be a `PyccelAstNode`. I would imagine that you would then have:
Dtype=int
Rank=0
Shape=()
Precision=default_precision ['int']
Order=None

But it might also make sense to have
Rank=1
Shape=(3,)

Comment (saniQQ) [reply pyccel/codegen/printing/ccode.py]: i think having the rank and shape 0 or `none` is more close to python and won't make a confusion to future devs

**Review by EmilyBourne — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

File: pyccel/codegen/printing/ccode.py (hunk 4 above)
Comment (yguclu) [reply pyccel/codegen/printing/ccode.py]: So I suppose this can be improved in a separate PR?
